### DISCUSSION TITLE
method textDocument/completion is not supported by any of the servers registered for the current buffer

### DISCUSSION THREAD

**Discussion by jotacosta-fb** [nvim-mini/mini.nvim#121] [Q&A] [answered]
I'm running completion module successfully. However, if I am editing a filetype that has servers but where those servers do not have completion, I keep getting this message:

> method textDocument/completion is not supported by any of the servers registered for the current buffer

This is correct, but how do I silence this? 

For example: when I'm editing a fish shell script with null-ls for formatting, I know I don't have completion. This makes it almost impossible to edit the script as it pop-up frequently.

**Comment by echasnovski** [comment, marked as answer]
Well, that shouldn't happen. I have only 'null-ls' formatting set up for markdown files and it works as expected - always uses fallback completion. Does this happen only with fish shell script files?

Could you please run `:lua require('mini.misc').put_text(vim.lsp.get_active_clients({ bufnr = 0 }))` in buffer where you are getting this behavior and post the output lines here? Don't forget to replace any absolute paths from output (like root directory, etc.).

**Reply by jotacosta-fb** [reply]
Just tested - it happens to me on json files too. I wonder if I miss something on the module configuration. I am using the default `require('mini.completion').setup {}`

**Reply by echasnovski** [reply]
Aha, thanks to this I was able to reproduce. The reason this doesn't happen in my setup is because I manually set function on attach of every LSP server that happen to have completion capability. This shouldn't be the case and I'll try to fix this tomorrow.

In the meantime here is what should work as (temporary or permanent) solution:
- Use this 'mini.completion' setup:
```lua
require('mini.completion').setup({
  lsp_completion = { source_func = 'omnifunc', auto_setup = false },
})
```
- Manually set appropriate omnifunc on attach every server you would like to have completion. [Here](https://github.com/echasnovski/nvim/blob/487ce206d88412db5577435ba956fcf5a19d6302/lua/ec/configs/nvim-lspconfig.lua#L11-L26) is how I do it (with [this usage](https://github.com/echasnovski/nvim/blob/487ce206d88412db5577435ba956fcf5a19d6302/lua/ec/configs/nvim-lspconfig.lua#L53)). I generally [recommend this approach](https://github.com/echasnovski/mini.nvim/blob/93e0f4b2e886813c65c94c5071f08cac6aacaed1/lua/mini/completion.lua#L62-L68).

**Reply by jotacosta-fb** [reply]
I have almost the same setup. Was super easy to just adopt your solution. 

Thanks!

**Comment by jotacosta-fb** [comment]
```
{ {
    _on_attach = <function 1>,
    attached_buffers = { true },
    cancel_request = <function 2>,
    commands = {},
    config = {
      cmd = { "nvim" },
      filetypes = { "python", "vim", "fish", "lua", "markdown", "yaml", "html", "json", "sh", "zsh" },
      flags = {
        debounce_text_changes = 250
      },
      get_language_id = <function 3>,
      name = "null-ls",
      on_attach = <function 4>,
      on_exit = <function 5>,
      on_init = <function 6>,
      root_dir = "/Users/jotacosta",
      settings = {}
    },
    handlers = {},
    id = 1,
    initialized = true,
    is_stopped = <function 7>,
    messages = {
      messages = {},
      name = "null-ls",
      progress = {},
      status = {}
    },
    name = "null-ls",
    notify = <function 8>,
    offset_encoding = "utf-16",
    request = <function 9>,
    request_sync = <function 10>,
    requests = {},
    rpc = {
      handle = {
        is_closing = <function 11>,
        kill = <function 12>
      },
      notify = <function 13>,
      pid = 5001,
      request = <function 14>
    },
    server_capabilities = {
      codeActionProvider = {
        resolveProvider = false
      },
      completionProvider = {
        allCommitCharacters = {},
        completionItem = {
          labelDetailsSupport = true
        },
        resolveProvider = false,
        triggerCharacters = { ".", ":", "-" }
      },
      documentFormattingProvider = true,
      documentRangeFormattingProvider = true,
      executeCommandProvider = true,
      hoverProvider = true,
      textDocumentSync = {
        change = 1,
        openClose = true,
        save = {
          includeText = true
        }
      }
    },
    stop = <function 15>,
    supports_method = <function 16>,
    workspaceFolders = <1>{ {
        name = "/Users/jotacosta",
        uri = "file:///Users/jotacosta"
      } },
    workspace_folders = <table 1>,
    <metatable> = {
      __index = <function 17>
    }
  } }
```

**Reply by echasnovski** [reply]
Yeah, it almost same as mine. Not an issue, though. See [this comment](https://github.com/echasnovski/mini.nvim/discussions/121#discussioncomment-3485410).

**Reply by echasnovski** [reply]
So, as far as I can tell, this output (and the one I have as well) tells that this particular LSP server *should* support completion, because there is a non-`nil` `server_capabilities.completionProvider` element. Here is the [source](https://github.com/neovim/neovim/issues/14090#issuecomment-1113956767) of supposedly canonical way of checking if this. The `completionProvider` is checked for `textDocument/completion` based on [this table entry](https://github.com/neovim/neovim/blob/09c6ce8c4e4c6415cca9b834539ed0df461373f6/runtime/lua/vim/lsp.lua#L58) from Neovim core.

So I am not sure what to do here, as at the moment it *feels* like an issue with 'null-ls', but it might as well be its intended behavior.
